ferry.go

```diff
@@ -310,35 +312,54 @@ func (f *Ferry) Run() {
 
 	go func() {
 		defer supportingServicesWg.Done()
-		handleError("throttler", f.Throttler.Run(ctx))
+		handleError("throttler", f.Throttler.Run(throttlerCtx))
 	}()
 
-	coreServicesWg := &sync.WaitGroup{}
-	coreServicesWg.Add(3)
+	binlogWg := &sync.WaitGroup{}
+	binlogWg.Add(2)
 
 	go func() {
-		defer coreServicesWg.Done()
-
-		f.BinlogStreamer.Run()
-		f.BinlogWriter.Stop()
+		defer binlogWg.Done()
+		f.BinlogWriter.Run(ctx)
 	}()
 
 	go func() {
-		defer coreServicesWg.Done()
-		f.BinlogWriter.Run()
+		defer binlogWg.Done()
+
+		f.BinlogStreamer.Run(ctx)
+		f.BinlogWriter.Stop()
 	}()
 
+	dataIteratorWg := &sync.WaitGroup{}
+	dataIteratorWg.Add(1)
+
 	go func() {
-		defer coreServicesWg.Done()
-		f.DataIterator.Run()
+		defer dataIteratorWg.Done()
+		f.DataIterator.Run(ctx)
 	}()
 
-	coreServicesWg.Wait()
+	dataIteratorWg.Wait()
 
+	f.logger.Info("data copy is complete, waiting for cutover")
+	f.OverallState = StateWaitingForCutover
+	f.waitUntilAutomaticCutoverIsTrue(ctx)
+
+	f.logger.Info("entering cutover phase, notifying caller that row copy is complete")
+	f.OverallState = StateCutover
+	f.notifyRowCopyComplete(ctx)
+
+	binlogWg.Wait()
+
+	f.logger.Info("ghostferry run is complete, shutting down auxiliary services")
```

Comment:
I wouldn't say "complete". An interrupted run is not complete.

Also "shutting down auxiliary services" is an implementation detail that i cannot see any user caring about

I would just say "ghostferry is shutting down".